src/libostree/ostree-sysroot-deploy.c

```diff
@@ -973,17 +978,88 @@ checksum_from_kernel_src (const char   *name,
+
+  int pipefds[2];
+  if (!g_unix_open_pipe (pipefds, FD_CLOEXEC, error))
     return FALSE;
-  if (syncfs (child_dfd) != 0)
-    return glnx_throw_errno_prefix (error, "syncfs(%s)", path);
+  glnx_fd_close int pipe_read = pipefds[0];
+  glnx_fd_close int pipe_write = pipefds[1];
+
+  pid_t pid = fork ();
+  if (pid < 0)
+    return glnx_throw_errno_prefix (error, "fork");
 
+  char c = '!';
+  if (pid == 0) /* Child watchdog/unfreezer process. */
+    {
+      if (daemon (0, 0) < 0)
+        err (1, "daemon");
+      int sigs[] = { SIGINT, SIGTERM };
+      for (guint i = 0; i < G_N_ELEMENTS (sigs); i++)
+        {
+          if (signal (sigs[i], SIG_IGN) == SIG_ERR)
+            err (1, "signal");
+        }
+      if (TEMP_FAILURE_RETRY (write (pipe_write, &c, sizeof (c))) < 0)
+        err (1, "write");
+      /* Wait for the parent to say it's going to freeze. */
+      (void) TEMP_FAILURE_RETRY (read (pipefds[0], &c, sizeof (c)));
+      /* Now we wait for the second message from the parent saying the freeze is
+       * complete. We have a 30 second timeout; if somehow the parent hasn't
+       * signaled completion, go ahead and unfreeze.
+       */
+      const int timeout_ms = 30000;
+      struct pollfd pfds[1];
+      pfds[0].fd = pipe_read;
+      pfds[0].events = POLLIN | POLLHUP;
+      int r = TEMP_FAILURE_RETRY (poll (pfds, 1, timeout_ms));
+      /* Do a thaw if we hit an error, or if the poll timed out */
+      if (r <= 0)
+        {
+          if (ioctl (rootfs_dfd, FITHAW, 0) != 0)
+            err (1, "FITHAW");
+          /* But if we got an error from poll, let's log it */
+          if (r < 0)
+            err (1, "poll");
+        }
+      exit (EXIT_SUCCESS);
+    }
+  else /* Parent process. */
+    {
+      /* Wait for the watchdog to say it's set up; mainly that it's
+       * masked SIGTERM successfully.
+       */
+      if (TEMP_FAILURE_RETRY (read (pipe_read, &c, sizeof (c))) < 0)
+        return glnx_throw_errno_prefix (error, "read(watchdog init)");
+      /* And tell the watchdog that we're ready to start */
+      if (TEMP_FAILURE_RETRY (write (pipe_write, &c, sizeof (c))) < 0)
+        return glnx_throw_errno_prefix (error, "write(watchdog start)");
+      /* Do a freeze/thaw cycle; TODO add a FIFREEZETHAW ioctl */
+      if (ioctl (rootfs_dfd, FIFREEZE, 0) != 0)
+        {
+          if (errno == EOPNOTSUPP)
+            {
+              /* Not supported?  OK, let's just do a syncfs */
+              if (syncfs (rootfs_dfd) != 0)
+                return glnx_throw_errno_prefix (error, "syncfs");
+              /* Write the completion, and return */
+              if (TEMP_FAILURE_RETRY (write (pipe_write, &c, sizeof (c))) < 0)
+                return glnx_throw_errno_prefix (error, "write(watchdog syncfs complete)");
+              return TRUE;
+            }
+          else
+            return glnx_throw_errno_prefix (error, "ioctl(FIFREEZE)");
+        }
+      if (ioctl (rootfs_dfd, FITHAW, 0) != 0)
```

Comment:
Maybe a shortcircuit here based on some env var so we can add an insttest that the watchdog kicks in as predicted?